Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,488 @@
+namespace Microsoft.ApplicationInsights.DependencyCollector.Implementation
+{
+    using System;
+    using System.Collections;
+    using System.Collections.Concurrent;
+    using System.Collections.Generic;
+    using System.Data;
+    using System.Data.SqlClient;
+    using System.Diagnostics;
+    using Microsoft.ApplicationInsights.DataContracts;
+    using Microsoft.ApplicationInsights.Extensibility;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation;
+    using Microsoft.ApplicationInsights.Extensibility.Implementation.Tracing;
+    using Microsoft.ApplicationInsights.Web.Implementation;
+
+    internal class SqlClientDiagnosticSourceListener : IObserver<KeyValuePair<string, object>>, IDisposable
+    {
+        // Event ids defined at: https://github.com/dotnet/corefx/blob/master/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlClientDiagnosticListenerExtensions.cs
+        public const string DiagnosticListenerName = "SqlClientDiagnosticListener";
+
+        public const string SqlBeforeExecuteCommand = SqlClientPrefix + "WriteCommandBefore";
+        public const string SqlAfterExecuteCommand = SqlClientPrefix + "WriteCommandAfter";
+        public const string SqlErrorExecuteCommand = SqlClientPrefix + "WriteCommandError";
+        
+        public const string SqlBeforeOpenConnection = SqlClientPrefix + "WriteConnectionOpenBefore";
+        public const string SqlAfterOpenConnection = SqlClientPrefix + "WriteConnectionOpenAfter";
+        public const string SqlErrorOpenConnection = SqlClientPrefix + "WriteConnectionOpenError";
+
+        public const string SqlBeforeCloseConnection = SqlClientPrefix + "WriteConnectionCloseBefore";
+        public const string SqlAfterCloseConnection = SqlClientPrefix + "WriteConnectionCloseAfter";
+        public const string SqlErrorCloseConnection = SqlClientPrefix + "WriteConnectionCloseError";
+
+        public const string SqlBeforeCommitTransaction = SqlClientPrefix + "WriteTransactionCommitBefore";
+        public const string SqlAfterCommitTransaction = SqlClientPrefix + "WriteTransactionCommitAfter";
+        public const string SqlErrorCommitTransaction = SqlClientPrefix + "WriteTransactionCommitError";
+
+        public const string SqlBeforeRollbackTransaction = SqlClientPrefix + "WriteTransactionRollbackBefore";
+        public const string SqlAfterRollbackTransaction = SqlClientPrefix + "WriteTransactionRollbackAfter";
+        public const string SqlErrorRollbackTransaction = SqlClientPrefix + "WriteTransactionRollbackError";
+
+        private const string SqlClientPrefix = "System.Data.SqlClient.";
+
+        private readonly TelemetryClient client;
+        private readonly SqlClientDiagnosticSourceSubscriber subscriber;
+
+        private readonly ConcurrentDictionary<Guid, long> operationStartTimestamps = new ConcurrentDictionary<Guid, long>();
```

Comment:
are there any cases including faulty ones when items will be added, but not removed: exceptions, DB unavailability, anything?

Reply:
@SergeyKanzhelev / @Dmitry-Matveev - please share your thoughts about the Dictionary usage here. I am not sure if its safe to assume callbacks would always come and dictionary never grows indefinitely. All other DependencyModules have some safety mechanism to clean Dictionary if something misbehaves.

Reply:
I would agree with Cijo on this one.

In the sample of SqlClient provided above, if [Statistics](https://github.com/dotnet/corefx/blob/83b606b5debad20536a581c425f66f4e9c396ed1/src/System.Data.SqlClient/src/System/Data/SqlClient/SqlCommand.cs#L811) call fails by any reason (e.g. NullRef on "." inside that method and further down the callstack), no End or Error is getting logged. Additionally, in the current implementation, `TryRemove` is not called in `finally`, so it's prone to the issues with the module initialization and runtime troubles where the code before it may fail.

I would also say that no cleanup adds a dependency on the internal implementation of the another module (our current design seem to rely on that dependency present state and the fact that it's known to us). This adds a bond between two modules that we are aware about on our end at the moment, but future owners of `SqlClient` or `DependencyCollector` might not be.

With the amount of the SQL calls enterprise apps do, this may grow pretty fast in case of any issues we haven't accounted for. At the same time, I understand that cleanup (if implemented) would need to make sure that telemetry items are still sent out correctly, so there may not be a simple immediate solution.
Can you please give it some thought and see if all of this make sense :) ?

Reply:
@Dmitry-Matveev Thanks for the feedback, I understand where you are coming from. 

My general thoughts on this are that, as engineers, we are constantly evaluating these types of trade-offs - In this case, balancing the extra cost and complexity of a more "defensive" approach against the less expensive, simpler, and more maintainable solution.

I was leaning towards the simpler side in this case after reviewing the SqlClient code and based on my experience using SqlClient - As you point out, that code is massively used, and I have never seen or heard of such a failure - Of course, that is not to say it can't ever happen, it just seemed very unlikely.

I can try to update the impl. Is the best approach to use one of _CacheBasedOperationHolder_ or _ObjectInstanceBasedOperationHolder_. If so, which one?

Reply:
If you are using id's as the key, then CacheBasedOperationHolder. The later, if you store the actual object itself in the dictionary.
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/FrameworkSqlProcessing.cs  This shows usage of cachebasedoperation holde.r